synthesis/synthesis.go

```diff
@@ -162,94 +154,23 @@ func findProblems(sequence string, problematicSequenceFuncs []func(string, chan
-	// using to solve synthesis fixing problems. The PRAGMA foreign_keys forces
-	// any reference to actually reference another object in the database, which
-	// reduces the quantity of potential errors.
-	// The other tables have basic create functions.
-	//
-	// Overall, the database is structured with 4 important tables:
-	// - SuggestedFix, which has suggested fixes for a sequence
-	// - Weights, which gives the weight of potential changes at any position
-	// - CodonBias, which caches the the bias of any codon to any other codon
-	// - History, which has the history of changes of a sequence
-	//
-	// Queries will generally take a suggestedFix, check the codon bias and weight
-	// of all potential changes within a range, and then pick appropriate changes,
-	// which are then added into the history.
-	createMemoryDbSQL := `
-	PRAGMA foreign_keys = ON;
-	CREATE TABLE codon (
-		codon TEXT PRIMARY KEY,
-		aa TEXT
-	);
-
-	CREATE TABLE seq (
-		pos INT PRIMARY KEY
-	);
-
-	-- Weights are set on a per position basis for codon harmonization at a later point
-	CREATE TABLE weights (
-		pos INTEGER REFERENCES seq(pos),
-		codon TEXT NOT NULL,
-		weight INTEGER,
-		FOREIGN KEY(codon) REFERENCES codon(codon)
-	);
-
-	CREATE TABLE codonbias (
-		gcbias TEXT CHECK(gcbias IN ('NA', 'GC', 'AT')),
-		fromcodon TEXT NOT NULL,
-		tocodon TEXT NOT NULL,
-		FOREIGN KEY(fromcodon) REFERENCES codon(codon),
-		FOREIGN KEY(tocodon) REFERENCES codon(codon)
-	);
-
-	CREATE TABLE suggestedfix (
-		id INTEGER PRIMARY KEY AUTOINCREMENT,
-		step INTEGER,
-		start INT NOT NULL,
-		end INT NOT NULL,
-		gcbias TEXT,
-		quantityfixes INTEGER,
-		suggestiontype TEXT,
-		FOREIGN KEY(start) REFERENCES seq(pos),
-		FOREIGN KEY(end) REFERENCES seq(pos)
-	);
-
-	CREATE TABLE history (
-                pos INTEGER,
-                codon TEXT NOT NULL,
-                step INT,
-                suggestedfix INT,
-                FOREIGN KEY(codon) REFERENCES codon(codon),
-                FOREIGN KEY(suggestedfix) REFERENCES suggestedfix(id),
-		FOREIGN KEY(pos) REFERENCES seq(pos)
-        );
-`
-	tx.MustExec(createMemoryDbSQL)
-	// Insert codons
-	weightTable := make(map[string]int)
-	codonInsert := `INSERT INTO codon(codon, aa) VALUES (?, ?)`
-	// First just insert the codons
-	for _, aminoAcid := range codontable.AminoAcids {
-		for _, codon := range aminoAcid.Codons {
-			tx.MustExec(codonInsert, codon.Triplet, aminoAcid.Letter)
-		}
-	}
-	// Then, add in GC biases
+	// Setup maps
+	// We have a historical map, a relative weight map, and a potential changes map.
+	historicalMap := make(map[int][]string)
+	weightMap := make(map[string]float64)
+	naBiasMap := make(map[string][]string)
+	gcBiasMap := make(map[string][]string)
+	atBiasMap := make(map[string][]string)
+
+	// Build historical maps and full amino acid weights
+	aminoAcidWeightTable := make(map[string]int)
 	for _, aminoAcid := range codontable.AminoAcids {
```

Comment:
These triple nested for loops need line by line comments to explain what's happening.